### ISSUE TITLE
Helm Charts for a Daemonset Deployment.

### ISSUE THREAD

**Issue by astub** [netdata/netdata#3727]
I've started writing a chart for running a netdata as a Daemonset, but I was wondering if anyone who already done this.

I have ran into many issues doing this in 1.9 k8 due to rbac issues and and mounting the host fs.

Anyone have any tips or something to share?

**Comment by ktsaou** [comment]
Hi,

this is really interesting. Did you manage to do it? I don't see why netdata should not work with a daemonset.

**Comment by ktsaou** [comment]
btw, daemonset docs: https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/

**Comment by hwinkel** [comment]
https://github.com/amazeeio/openshift-netdata

**Comment by paulfantom** [comment]
@hwinkel I wonder when someone will write a netdata operator for openshift :smile:


@astub helm charts should be stored in https://github.com/helm/charts repository and guys from that repo should be able to help in solving issues regarding charts.

**Comment by majodev** [comment]
Hi,

I took me quite a time to get netdata to play nice in your k8s infrastructure. For reference, here's the daemonset I'm using: https://gist.github.com/majodev/cdacbf1a97a20d7404d40b22d12f3d90

Major hurdles:
* host volume mount propagation must be enabled (k8s >=1.10)
* netdata group is added to root group within the docker container (to access host mounted volumes)
* cgroup network lookup is disabled (`echo`), else major performance penalty during startup
* `securityContext`, `capabilities` and `PGID` must be set
* apparmor must be disabled
* cgroup name resolution order must be changed, also exclude `var-*.scope` system slices

**Comment by ktsaou** [comment]
@majodev 	this is very nice! Thank you for sharing your findings.

> * cgroup network lookup is disabled (`echo`), else major performance penalty during startup

I am not sure I understand this. Can you elaborate a bit?

> * apparmor must be disabled

hm... we could probably provide aparmor config for netdata?

> * cgroup name resolution order must be changed

I am not sure I understand this either.

> * also exclude `var-*.scope` system slices

We could exclude these by default in netdata. Right?

Generally, do you believe we should do something to make the whole thing simpler? I would love to work to provide a smoother experience with k8s.

**Comment by varyumin** [comment]
I have tried to write the helm chart. This is the helm chart fully close my cases. Maybe it is helpful to somebody and I open to discussion )))

**Comment by gmosx** [comment]
> I have tried to write the helm chart.

Looks interesting, thanks!

**Comment by cakrit** [comment]
Hi @varyumin, it's excellent timing actually, we just got into this seriously. I have a few questions for your right from the start and more will follow as we learn more about this and continue our testing:
- We will create a project under netdata so we can provide our official helm chart. Of course it would be great if we can start from your code as a baseline. Are you ok with that? You'll get full credit of course!

- I noticed the nginx ingress part after the first installation.  Is there a dependency on having nginx-ingress installed first? What would `netdata.k8s.local` resolve to? I'm a newbie and just did a helm install on minikube and fell into the following path:
```
1. Get the application URL by running these commands:
{{- if .Values.ingress.enabled }}
{{- range .Values.ingress.hosts }}
  http{{ if $.Values.ingress.tls }}s{{ end }}://{{ . }}{{ $.Values.ingress.path }}
```
But the URL didn't work and I added a service of type `NodePort`to expose the UI.

I will do comparisons with the work @majodev has done as well, but we'll also be doing a lot of testing before making anything official. 

@hwinkel I completely missed your comment on openshift. It's a hell of a lot of work and deserves much more than just a link! We'll try that one too of course.

**Comment by cakrit** [comment]
@varyumin : I figured out the ingress part, after reading some more docs. I have to say, the chart you created is pretty awesome! I learned a lot from it. Using the ingress controller, having a Statefulset for the master and Daemonsets for the slaves is clearly the way to go.  Going more deeply now, to get netdata to behave properly in this environment.

**Comment by varyumin** [comment]
Hi @cakrit. If you don't have or don't use the ingress controller, you can set 
```
ingress:
  enabled: false
```
but you need to get access to web ui, therefore you should use nodePort in service (I rewrote a little service ). You can set:
#### If you have to get random NodePort
```
service:
  type: NodePort
```
```
$   kubectl describe service netdata

Name:                     netdata
Namespace:                default
Labels:                   app=netdata
                          chart=netdata-0.1.0
                          heritage=Tiller
                          release=sad-elk
                          role=master
Annotations:              <none>
Selector:                 app=netdata,release=sad-elk,role=master
Type:                     NodePort
IP:                       10.105.199.23
Port:                     http  19999/TCP
TargetPort:               http/TCP
NodePort:                 http 32111/TCP
Endpoints:                10.105.6.241:19999
Session Affinity:         None
External Traffic Policy:  Cluster
Events:                   <none>
```
in my example you can see, I got port **NodePort: http 32111/TCP**
#### If you have to get personal port to NodePort
```
service:
  type: NodePort
  nodePort: 32555
```

```
Name:                     netdata
Namespace:                default
Labels:                   app=netdata
                          chart=netdata-0.1.0
                          heritage=Tiller
                          release=joking-tuatara
                          role=master
Annotations:              <none>
Selector:                 app=netdata,release=joking-tuatara,role=master
Type:                     NodePort
IP:                       10.105.235.96
Port:                     http  19999/TCP
TargetPort:               http/TCP
NodePort:                 http  32555/TCP
Endpoints:                
Session Affinity:         None
External Traffic Policy:  Cluster
Events:                   <none>

```
**NodePort: http  32555/TCP**

**Comment by cakrit** [comment]
Thanks a lot, but the ingress controller is clearly the better way to do it. I extended the chart with Clusterrole etc. and it works fine in minikube, I got to query the API to translate the cgroups. Today I started with Google Kubernetes Engine and it looks like I will need to do more changes. 

We're keeping the helmchart as a private project right now, until it works reasonably well for most cases and until I have your consent to publish it (since I based it on your work). Would you like me to invite you and add you as code owner of that project?

**Comment by varyumin** [comment]
Hi @cakrit. 
` Would you like me to invite you and add you as code owner of that project?`
Yes, I do. Thx

**Comment by varyumin** [comment]
I could write "Kubernetes operator" for netdata. If you need.
operator-sdk:
https://github.com/operator-framework/operator-sdk

**Comment by gmosx** [comment]
> I could write "Kubernetes operator" for netdata.

That would be fantastic!

**Comment by cakrit** [comment]
I made https://github.com/netdata/helmchart/ public and added you as an owner for now, I'll put the CODEOWNERS file later. You'll see that the current chart pulls a test image created for PR #5576 , so that we show the correct cgroup names. There are several other additions/changes as well and I'm going to be writing a lot on the README about how it all works. 

My main problem right now is that with the persistent volume claims I added to persist the metric and alarm DBs, GKE is doing something weird. It apparently binds the volumes just fine, but when netdata starts, the filesystem is a mess (first indication was that /proc/1/sched was missing and then it couldn't actually read `/var/lib/netdata`). So there's definitely something wrong there. It does work in minikube though. I'd appreciate your help @varyumin, we can continue the discussion here and in PRs on that project.

An operator would really be awesome, we'll definitely be heading there. Of course I'd first like to have a good, stable, supported sample configuration that we can announce, to make people's lives easier to get started. I mean, you'll see that I added to this Chart a sample config for health notifications using email and our Docker image doesn't enable sendmail, so it's of little use right now. So I'd say we first need to stabilize the helm chart and get feedback from the community to move it out of beta, and then look into the operator.

**Comment by cakrit** [comment]
The issue with the persistent volumes in GKE has been fixed and I'll be closing this issue with #5627, which adds a link to the helmchart installation instructions. You'll see on the project https://github.com/netdata/helmchart a first issue on additional TODOs. Of course we'd appreciate contributions and requests from anyone who will use it.
